tensorflow/c/experimental/profiler/profiler.cc

```diff
@@ -0,0 +1,225 @@
+  VALIDATE_MEMBER(TF_ProfilerRegistrationParams, params, destroy_profiler_fns);
+  return Status::OK();
+}
+
+Status ValidateTPProfiler(const TP_Profiler& profiler) {
+  VALIDATE_STRUCT_SIZE(TP_Profiler, profiler, TP_PROFILER_STRUCT_SIZE);
+  VALIDATE_MEMBER(TP_Profiler, profiler, type);
+  TF_RETURN_IF_ERROR(ValidateDeviceType(profiler.type));
+  return Status::OK();
+}
+
+Status ValidateTPProfilerFns(const TP_ProfilerFns& profiler_fns) {
+  VALIDATE_STRUCT_SIZE(TP_ProfilerFns, profiler_fns,
+                       TF_PROFILER_FNS_STRUCT_SIZE);
+  VALIDATE_MEMBER(TP_ProfilerFns, profiler_fns, start);
+  VALIDATE_MEMBER(TP_ProfilerFns, profiler_fns, stop);
+  VALIDATE_MEMBER(TP_ProfilerFns, profiler_fns, collect_data_xspace);
+  return Status::OK();
+}
+
+class PluggableProfiler : public tensorflow::profiler::ProfilerInterface {
+ public:
+  // The caller must have validated profiler_fns and profiler.
+  static std::unique_ptr<tensorflow::profiler::ProfilerInterface>
+  CreatePluggableProfiler(const ProfileOptions& options, TP_Profiler profiler,
+                          TP_ProfilerFns profiler_fns) {
+    if (options.device_tracer_level() == 0) {
+      return nullptr;
+    }
+    if (options.device_type() != ProfileOptions::UNSPECIFIED &&
+        options.device_type() != ProfileOptions::PLUGGABLE_DEVICE) {
+      return nullptr;
+    }
+    return absl::WrapUnique(new PluggableProfiler(profiler_fns, profiler));
+  }
+
+  Status Start() {
+    std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());
+    profiler_fns_.start(&profiler_, c_status.get());
+    return tensorflow::StatusFromTF_Status(c_status.get());
+  }
+
+  Status Stop() {
+    std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());
+    profiler_fns_.stop(&profiler_, c_status.get());
+    return tensorflow::StatusFromTF_Status(c_status.get());
+  }
+
+  Status CollectData(XSpace* space) {
+    std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());
+    // Get size of buffer required for Plugin to serialize XSpace ito.
+    size_t size_in_bytes;
+    profiler_fns_.collect_data_xspace(&profiler_, /*buffer=*/nullptr,
+                                      &size_in_bytes, c_status.get());
+
+    // Prepare an appropriately sized buffer.
+    if (size_in_bytes > 0) {
+      std::vector<uint8_t> buffer(size_in_bytes);
+      profiler_fns_.collect_data_xspace(&profiler_, buffer.data(),
+                                        &size_in_bytes, c_status.get());
+      // Deserialize XSpace from the buffer and return it.
+      XSpace plugin_space;
+      plugin_space.ParseFromArray(buffer.data(), buffer.size());
+      for (XPlane& plugin_plane : *plugin_space.mutable_planes()) {
+        XPlane* plane = space->add_planes();
+        plane->Swap(&plugin_plane);
+      }
+    }
+    return tensorflow::StatusFromTF_Status(c_status.get());
+  }
+
+  Status CollectData(RunMetadata* run_metadata) {
```

Comment:
This function can be removed, RunMetadata overload of CollectData has been removed from ProfilerInterface.

Reply:
Done